File: plugins/de-groot/models/box.py
from database.db import db
from flask_mongoengine import BaseQuerySet
import datetime
from slugify import slugify, Slugify, UniqueSlugify
from models.category import Category
from flask_mongoengine.wtf import model_form

# id = lambda: str(uuid.uuid4())


class Box(db.Document):
    sort = db.IntField(required=False, default=0)
    tenant_id = db.StringField(required=False, default=None)
    tenant_name = db.StringField(required=False)
    sku = db.StringField(required=False, default='')
    name = db.StringField(required=True, unique=True)
    display_name = db.ListField(required=False, default=[])
    system_key = db.StringField(required=False, default="")
    slug = db.StringField(required=False, unique=True)
    description = db.StringField(required=False, default='')
    media = db.ListField()
    sqm = db.BooleanField(required=True, unique=False, default=False)
    appendage = db.BooleanField(required=True, unique=False, default=False)
    calculation_type = db.StringField(default='')
    published = db.BooleanField(default=True)
    input_type = db.StringField(default='')
    incremental = db.BooleanField(required=True, default=False)
    select_limit = db.IntField(required=False, default=0)
    option_limit = db.IntField(required=False, default=0)
    shareable = db.BooleanField(default=True)
    start_cost = db.IntField(required=False)
    created_at = db.DateTimeField(default=datetime.datetime.now, required=True)
    categories = db.ListField(db.ReferenceField(Category))
    additional = db.DictField(required=False, default={})
    checked = db.BooleanField(default=False)

    configure = db.DictField(required=False,default={})

    meta = {
        'indexes': ['-created_at', 'slug', 'name', 'categories'],
        'ordering': ['-created_at', 'name'],
        'collection': 'boxes',
        'queryset_class': BaseQuerySet
    }

    def __setattr__(self, key, value):
        super(Box, self).__setattr__(key, value)
        if key == 'name':
            self.slug = slugify(self.name, to_lower=True)

File: microservices/categories/categories/Models/Box.js
const mongoose = require('mongoose');
const slugify = require('slugify');

// Define the Box schema
const BoxSchema = new mongoose.Schema({
  sort: { type: Number, default: 0 },
  tenant_id: { type: String, default: null },
  tenant_name: { type: String, required: false },
  sku: { type: String, default: '' },
  name: { type: String, required: true, unique: true },
  display_name: { type: [String], default: [] },
  system_key: { type: String, default: '' },
  slug: { type: String, unique: true },
  description: { type: String, default: '' },
  media: { type: [String], default: [] },
  sqm: { type: Boolean, required: true, default: false },
  appendage: { type: Boolean, required: true, default: false },
  calculation_type: { type: String, default: '' },
  published: { type: Boolean, default: true },
  input_type: { type: String, default: '' },
  incremental: { type: Boolean, required: true, default: false },
  select_limit: { type: Number, default: 0 },
  option_limit: { type: Number, default: 0 },
  shareable: { type: Boolean, default: true },
  start_cost: { type: Number, default: 0 },
  created_at: { type: Date, default: Date.now, required: true },
  categories: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Category' }],
  additional: { type: Map, of: String, default: {} },
}, {
  versionKey: false // You should be aware of the outcome after set to false
});

// Add an index to the schema (equivalent to 'meta' in MongoEngine)
BoxSchema.index({ created_at: -1 });
BoxSchema.index({ slug: 1 });
BoxSchema.index({ name: 1 });
BoxSchema.index({ categories: 1 });

// Pre-save hook to auto-generate slug from the name field
BoxSchema.pre('save', function (next) {
  if (this.isModified('name')) {
    this.slug = slugify(this.name, { lower: true });
    this.system_key = slugify(this.name, { lower: true });
  }
  next();
});

// Set the default sorting (equivalent to 'ordering' in MongoEngine)
BoxSchema.set('timestamps', true);

// Create the Box model from the schema
const Box = mongoose.model('Box', BoxSchema);

module.exports = Box;
